api/service/explorer/storage.go

```diff
@@ -69,78 +103,239 @@ func (storage *Storage) Init(ip, port string) {
 		WriteBuffer:            0,
 		Filter:                 filter.NewBloomFilter(10),
 	}
-	if storage.db, err = leveldb.OpenFile(dbFileName, options); err != nil {
+	db, err := leveldb.OpenFile(dbFileName, options)
+	if err != nil {
 		utils.Logger().Error().Err(err).Msg("Failed to create new database")
+		return nil, err
+	}
+	return db, nil
+}
+
+func (storage *Storage) Start() {
+	go storage.loop()
+}
+
+func (storage *Storage) Close() {
+	storage.closeC <- struct{}{}
+	<-storage.closeC
+}
+
+func (storage *Storage) loop() {
+	defer func() {
+		if err := storage.flushLocked(); err != nil {
+			utils.Logger().Error().Err(err).Msg("[explorer] failed to flush at close")
+		}
+		close(storage.closeC)
+	}()
+
+	var lastFlush = time.Now()
+
+	for {
+		// flush when reached threshold
+		storage.lock.RLock()
+		numDirty := len(storage.dirtyBNs)
+		storage.lock.RUnlock()
+		if numDirty >= flushNumberThreshold || time.Since(lastFlush) >= flushTimeThreshold {
+			lastFlush = time.Now()
+			if err := storage.flushLocked(); err != nil {
+				utils.Logger().Error().Err(err).Msg("[explorer] failed to flush")
+			}
+		}
+		// prioritize processing new blocks and close C
+		select {
+		case b := <-storage.newBlockC:
+			if err := storage.handleBlock(b); err != nil {
+				utils.Logger().Error().Err(err).Msg("[explorer] handle new block result failed")
+			}
+			continue
+		case <-storage.closeC:
+			return
+		default:
+		}
+		// handles catchupBlock
+		select {
+		case b := <-storage.newBlockC:
```

Comment:
why this newBlockC logic is repeated here?

Reply:
Here is the trick to prioritize handling `newBlockC` over `catchupBlockC`. This is not very clean though. Do you have another idea to prioritizing handling a channel over another?

Reply:
If this is to prioritize processing new block, why do we need the line from 145-154?

Reply:
Line 145-154 is to prioritze processing new block. I don't get it. Select case in golang does not have a priority on the competing channels. And this is the hack to achieve this.

Reply:
but why do we need to repeat the same thing again in the last select here? 157 - 160?

Reply:
Because the loop should be both waiting on newBlockC as well as catchupBlockC. If we do not repeat the same here, the loop will not continue if a block is sent to newBlockC, but not catchupBlockC.

Reply:
https://stackoverflow.com/questions/11117382/priority-in-go-select-statement-workaround

Reply:
looks good. thanks.